src/mpi/coll/transports/generic/transport.c

```diff
@@ -0,0 +1,1243 @@
+                                       int tag,
+                                       MPIR_COLL_GENERIC_comm_t * comm,
+                                       uint64_t flags,
+                                       MPIR_COLL_GENERIC_sched_t * sched, int n_invtcs, int *invtcs)
+{
+    int iscontig;
+    size_t type_size, out_extent, lower_bound;
+    MPIR_COLL_GENERIC_vtx_t *vtxp;
+
+    /* assign a vertex */
+    int vtx_id = MPIR_COLL_GENERIC_get_new_vtx(sched, &vtxp);
+    sched->tag = tag;
+    vtxp->kind = MPIR_COLL_GENERIC_KIND_RECV_REDUCE;
+    MPIR_COLL_GENERIC_init_vtx(sched, vtxp, vtx_id);
+    MPIR_COLL_GENERIC_add_vtx_dependencies(sched, vtx_id, n_invtcs, invtcs);
+
+    /* store the arguments */
+    MPIR_COLL_GENERIC_dtinfo(datatype, &iscontig, &type_size, &out_extent, &lower_bound);
+    vtxp->nbargs.recv_reduce.inbuf = MPIR_COLL_GENERIC_allocate_mem(count * out_extent);
+    vtxp->nbargs.recv_reduce.inoutbuf = buf;
+    vtxp->nbargs.recv_reduce.count = count;
+    vtxp->nbargs.recv_reduce.datatype = datatype;
+    vtxp->nbargs.recv_reduce.op = op;
+    vtxp->nbargs.recv_reduce.source = source;
+    vtxp->nbargs.recv_reduce.comm = comm;
+    vtxp->nbargs.recv_reduce.vtxp = vtxp;
+    vtxp->nbargs.recv_reduce.done = 0;
+    vtxp->nbargs.recv_reduce.flags = flags;
+
+    MPL_DBG_MSG_FMT(MPIR_DBG_COLL,VERBOSE,(MPL_DBG_FDEST,"TSP(generic) : sched [%d] [recv_reduce]\n", vtx_id));
+
+    return vtx_id;
+}
+
+/* Transport function to get rank in communicator */
+int MPIR_COLL_GENERIC_rank(MPIR_COLL_GENERIC_comm_t * comm)
+{
+    return comm->mpid_comm->rank;
+}
+
+/* Transport function to get size of the communicator */
+int MPIR_COLL_GENERIC_size(MPIR_COLL_GENERIC_comm_t * comm)
+{
+    return comm->mpid_comm->local_size;
+}
+
+/* Transport function to schedule a local reduce */
+int MPIR_COLL_GENERIC_reduce_local(const void *inbuf,
+                                        void *inoutbuf,
+                                        int count,
+                                        MPIR_COLL_GENERIC_dt_t datatype,
+                                        MPIR_COLL_GENERIC_op_t operation,
+                                        uint64_t flags,
+                                        MPIR_COLL_GENERIC_sched_t * sched, int n_invtcs, int *invtcs)
+{
+    MPIR_COLL_GENERIC_vtx_t *vtxp;
+    int vtx_id = MPIR_COLL_GENERIC_get_new_vtx(sched, &vtxp);
+
+    /* assign a new vertex */
+    vtxp->kind = MPIR_COLL_GENERIC_KIND_REDUCE_LOCAL;
+    vtxp->state = MPIR_COLL_GENERIC_STATE_INIT;
+    MPIR_COLL_GENERIC_init_vtx(sched, vtxp, vtx_id);
+    MPIR_COLL_GENERIC_add_vtx_dependencies(sched, vtx_id, n_invtcs, invtcs);
+
+    /* store the arguments */
+    vtxp->nbargs.reduce_local.inbuf = inbuf;
+    vtxp->nbargs.reduce_local.inoutbuf = inoutbuf;
+    vtxp->nbargs.reduce_local.count = count;
+    vtxp->nbargs.reduce_local.dt = datatype;
+    vtxp->nbargs.reduce_local.op = operation;
+    vtxp->nbargs.reduce_local.flags = flags;
```

Comment:
In the implementation, the array of vertices is already allocated. _get_new_vtx is just getting an index into the array. So, I think _get_new_vtx needs to be thread-safe. I haven't thought of thready safety at all. May be we should discuss it?

Reply:
Ok, if all it's getting is an index, then maybe it's ok since there wouldn't be two threads modifying the same memory address.

As for thread safety for the function itself, I think it just needs a mutex or perhaps an internal compare and swap.